[SuperTextField][Android][iOS] Disables text field caret blink while dragging caret (Resolves #1705)

rutvik110 opened the pull request:
[SuperTextField][Android][ios] Disables text field caret blink while dragging caret (Resolves #1705)

### **Issue:** 

Text field's caret keeps blinking while caret's being dragged within the text field. On both Android and iOS, in the native apps, the caret stops blinking as long as the user's dragging the caret. Same is expected from super text field on android and ios.

`SuperTextField` behaviour on Android and iOS:

https://github.com/superlistapp/super_editor/assets/65209850/6465ba53-134a-4d05-aeec-3107463c7bee


### Solution: 

This PR updates both `SuperAndroidTextField` and `SuperIOSTextField` and their necessary internal components to match the native behaviour for caret blinking when user's dragging the caret within text field.  As long as the drag's is happening, caret blinking is disabled, and once the drag ends, caret blinking is enabled for the text field.

With fix, `SuperTextField` behaviour on Android and iOS:

https://github.com/superlistapp/super_editor/assets/65209850/3b08287a-e71f-4791-b64d-03ca513ebcee


### Remarks: 

* `SuperAndroidTextField` and `SuperAndroidTextField` are updated to create an internal `BlinkController` that can be passed down to their respective `TextLayoutCaret` that displays and controlls the text field caret. 
* Fixes a bug in the `TextLayoutCaret`'s by updating the `createBlinkController` method to return the passed `BlinkController` if present instead of returning it's own internal `blinkController` variable.
* `AndroidTextFieldTouchInteractor` and `IOSEditingControls` are updated to receive the internal `BlinkController` from text field which is used to control blinking of caret on drag start/end within their drag interaction handlers.
* `AndroidTextFieldTouchInteractor` and `IOSTextFieldTouchInteractor` are updated to receive the internal `BlinkController` from text field which is used to control blinking of caret on drag start/end within their drag interaction handlers.

matthew-carroll commented on super_editor/lib/src/super_textfield/android/_editing_controls.dart:
```diff
@@ -298,6 +302,7 @@ class _AndroidEditingOverlayControlsState extends State<AndroidEditingOverlayCon
     setState(() {
       _localDragOffset = _localDragOffset! + details.delta;
       widget.editingController.showMagnifier(_localDragOffset!);
+      widget.blinkController.stopBlinking();
```
This is calling `stopBlinking()` 60 times per second while dragging, which seems unnecessary.

matthew-carroll commented on super_editor/lib/src/super_textfield/android/_editing_controls.dart:
```diff
@@ -50,6 +51,9 @@ class AndroidEditingOverlayControls extends StatefulWidget {
   final TextScrollController textScrollController;
 
+  /// Text field caret blink controller.
+  final BlinkController blinkController;
```
Given that `AndroidEditingOverlayController` is the controller that links this overlay UI to the text field, wouldn't that be the correct place to introduce blinking start/stop controls?

rutvik110 replied:
yah, that's right.  Will update this to include the `blinkController` as part of `AndroidEditingOverlayController`.

matthew-carroll commented on super_editor/lib/src/super_textfield/android/_user_interaction.dart:
```diff
@@ -77,6 +78,9 @@ class AndroidTextFieldTouchInteractor extends StatefulWidget {
   final TextScrollController textScrollController;
 
+  /// Text field caret blink controller.
+  final BlinkController blinkController;
```
Continuing with my comment from the overlay - we already have a controller that shares visual updates between the handles and the text field. It's `AndroidEditingOverlayController`. Shouldn't the blinking APIs be added there?

matthew-carroll commented on super_editor/lib/src/super_textfield/android/_user_interaction.dart:
```diff
@@ -335,6 +339,7 @@ class AndroidTextFieldTouchInteractorState extends State<AndroidTextFieldTouchIn
       );
 
       widget.editingOverlayController.showMagnifier(_globalDragOffset!);
+      widget.blinkController.stopBlinking();
```
This one also gets called 60 times per second. That seems unnecessary.

matthew-carroll commented on super_editor/lib/src/super_textfield/android/_user_interaction.dart:
```diff
@@ -359,6 +364,7 @@ class AndroidTextFieldTouchInteractorState extends State<AndroidTextFieldTouchIn
     setState(() {
       _isDraggingCaret = false;
       widget.editingOverlayController.hideMagnifier();
+      widget.blinkController.startBlinking();
```
Is it possible that the selection is expanded here? If so, what happens when you call start blinking with an expanded selection?

rutvik110 replied:
Technically, it shouldn't affect anything. `start/stop` on `blinkController` is mainly responsible for starting/stoping it's ticker that's used to flash the caret, if one's not being shown as in case of expanded selection, then it shouldn't have any effect. Also, tested it out and didn't saw any side effects of calling those here.

matthew-carroll replied:
Please keep in mind the ability for future developers to make changes to code. When you say "Technically, it shouldn't affect anything", it sounds like what you're saying is "This behavior isn't relevant for expanded selections and shouldn't be called in that case, but it doesn't break anything if we do".

The problem is, how does everyone else know this? The default assumption when reading all code, is that the code you're reading exists for a reason. Therefore, other developers may be hesitant to take action that removes `startBlinking()` for expanded selections. Moreover, developers might avoid making convenient changes for expanded selections because they think that their changes need to continue working with the `startBlinking()` line.

It's generally harmful to let code exist where it shouldn't simply because that code doesn't currently break something.

rutvik110 replied:
That's right. Updated to call start/stop blinking only when the selection is collapsed.

matthew-carroll commented on super_editor/lib/src/super_textfield/android/android_textfield.dart:
```diff
@@ -628,6 +642,7 @@ class SuperAndroidTextFieldState extends State<SuperAndroidTextField>
                 ? _textEditingController.selection.extent
                 : null,
             blinkTimingMode: widget.blinkTimingMode,
+            blinkController: _blinkController,
```
What does it mean to pass a blink controller and pass a blink timing mode?

rutvik110 replied:
ah well, the `blinkTimingMode` is used by `TextLayoutCaret` to create a ticker/timer based `blinkController` but as we're creating the controller above, it shouldn't matter if we pass it or not. But, `TextLayoutCaret` may've other use of `blinkTimingMode`, so I'm gonna keep it as it's.

matthew-carroll replied:
> But, TextLayoutCaret may've other use of blinkTimingMode,

This isn't something to be left to assumptions. Does it make other use of it, or not?

> ah well, the blinkTimingMode is used by TextLayoutCaret to create a ticker/timer based blinkController but as we're creating the controller above, it shouldn't matter if we pass it or not

For the same reason as my other comment on this PR, this perspective is generally unacceptable. We don't want to litter incorrect-but-not-breaking code just because we can. We want to maintain a codebase of intentional behaviors that do exactly what's intended.

> so I'm gonna keep it as it's.

Based on the rationale in your comment, this is probably the wrong conclusion.

rutvik110 replied:
> This isn't something to be left to assumptions. Does it make other use of it, or not?

It doesn't atm. 

> For the same reason as my other comment on this PR, this perspective is generally unacceptable. We don't want to litter incorrect-but-not-breaking code just because we can. We want to maintain a codebase of intentional behaviors that do exactly what's intended.

That's true but `TextLayoutCaret` sets `blinkTimingMode` default value to be ticker based. If the user happens to change this to timer based, and we didn't passed the user defined mode to `blinkTimingMode` in this case, then we would eventually have a false value for `blinkTimingMode` within `TextLayoutCaret`. Even if it self doesn't use it for anything other than creating the `blinkController` atm but if something else in the tree has to access it's `blinkTimingMode`, then it would work with a false value eventually. Based on this, shouldn't keeping it is the right choice here?

rutvik110 replied:
@matthew-carroll Pinging in case you missed this.

matthew-carroll replied:
I couldn't follow your explanation above. First, what specific API are you proposing? Second, what's a specific situation that requires the API you're proposing? Third, if you're proposing an API that takes a timing mode and a controller, how do those things not conflict with each other?

rutvik110 replied:
> First, what specific API are you proposing? Second, what's a specific situation that requires the API you're proposing? Third, if you're proposing an API that takes a timing mode and a controller, how do those things not conflict with each other?

I think this whole discussion is getting dragged by hypothetical what if questions from my side. For the moment, from our discussion and as I see it now, there's no need to pass the `BlinkTimingMode` here if we're already passing the controller. If there's any need in the future for passing `BlinkTimingMode` in presence of controller, we'll know it. But right now, this doesn't affect anything, so we're good.

Based on this, I'm removing the passed `blinkTimingMode` from here for now. Does that sound good?

matthew-carroll replied:
Sure. I can take another look after you try that.

rutvik110 replied:
Pushed the updates. Let me know if we're good here.

matthew-carroll reviewed (commented):
I left a number of comments on the Android approach. They probably apply similarly to the iOS approach. Also, you're missing tests.

rutvik110 commented:
@matthew-carroll Moved the caret start/stop controlls within the `editingController` for respective platforms. And also updated to signal start or stop blinking only if the caret has either stoped blinking or is blinking.  Tests are also added to ensure caret doesn't blink when dragging on android and ios, and when dragging the collapsed handle on android.

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_caret_test.dart:
```diff
@@ -166,6 +167,60 @@ void main() {
+      await tester.dragCaretByDistanceInSuperTextField(const Offset(100, 100));
+
+      // Ensure caret is visible.
+      expect(_isCaretVisible(tester), true);
```
I have two questions about this check. Did you make sure to re-enable caret animations for this test? If not, the caret will probably never blink.

Second, does `_isCaretVisible()` return `false` when the caret opacity is zero?

Third, did you check to make sure this test fails without your changes?

rutvik110 replied:
> Did you make sure to re-enable caret animations for this test? If not, the caret will probably never blink.

If you're referring to this check within some of the above tests, then it's not present rn in this added tests. I assumed the caret blinking was on by default. I'll update the tests with this.
```dart
// Configure BlinkController to animate, otherwise it won't blink
BlinkController.indeterminateAnimationsEnabled = true;
```

> Second, does _isCaretVisible() return false when the caret opacity is zero?

yes, `_isCaretVisible` explicitly checks for the opacity to ensure caret's visible and based on existing tests, it does.

> Third, did you check to make sure this test fails without your changes?

yes

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_caret_test.dart:
```diff
@@ -166,6 +167,60 @@ void main() {
       expect(caret.style.borderRadius, caretStyle.borderRadius);
     });
+
+    testWidgetsOnMobile("does not blink while dragging", (tester) async {
```
To clarify how this test differs from the one below it, please add the ending "does not blink while dragging the caret"

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_caret_test.dart:
```diff
@@ -196,3 +251,29 @@ bool _isCaretVisible(WidgetTester tester) {
 }
+
+/// Pumps a [SuperTextField] experience.
+Future<void> _pumpTestApp(
```
A few things here. First, higher level methods should appear above lower level methods. For example, pumping a whole app is higher level than pumping just a scaffold.

Second, why doesn't this method re-use the existing `_buildScaffold()`? You seem to have used the same root widget tree.

Third, this test name introduces confusion. It begs the question "why isn't this method used for all tests in this suite if it's really a "test app". If this method pumps something more specific, then it should be named accordingly.

Fourth, why not just pump the `SuperTextField` that you want directly in the tests like all other tests in this suite? It's only two tests, there's no special configuration, and that would retain consistency with everything else in this file.

rutvik110 replied:
This was based on an earlier implementation of the tests but  is redundant now. Removed it.

rutvik110 commented on super_editor/lib/src/super_textfield/android/_editing_controls.dart:
```diff
@@ -374,6 +379,10 @@ class _AndroidEditingOverlayControlsState extends State<AndroidEditingOverlayCon
       _isDraggingExtent = false;
       widget.editingController.hideMagnifier();
 
+      if (widget.editingController.textController.selection.isCollapsed) {
```
ah right, fixed it.

rutvik110 commented on super_editor/lib/src/super_textfield/android/_editing_controls.dart:
```diff
@@ -305,6 +305,11 @@ class _AndroidEditingOverlayControlsState extends State<AndroidEditingOverlayCon
       _localDragOffset = _localDragOffset! + details.delta;
       widget.editingController.showMagnifier(_localDragOffset!);
+
+      if (widget.editingController.textController.selection.isCollapsed) {
```
I'll check on this.

rutvik110 replied:
Based on our discussion, I've moved this call to pan start event callbacks for both handle and user interaction handlers on text field.

I noticed that overlay controllers `IOSEditingControls` and `AndroidEditingOverlayControls` were missing a `onPanStart` method so I updated them include a `onHandleDragStart()` method that's called within the respective `onPanStart` methods passed down to created handles.  

`onHandleDragStart` is responsible for calling common drag start events like hiding the toolbar, etc.

`AndroidEditingOverlayControls` methods  `_onBasePanStart`, `_onExtentPanStart` and `_onCollapsedPanStart` are updated to use `onHandleDragStart`.

And similarly `IOSEditingControls` methods `_onBasePanStart` and `_onExtentPanStart` are updated as well.

Beside moving the blink stop call within `onHandleDragStart`, some of the common logic that was shared across this pan start methods is moved within it as well.

rutvik110 commented on super_editor/test/super_textfield/super_textfield_caret_test.dart:
```diff
@@ -166,6 +167,60 @@ void main() {
+      // Ensure caret is visible.
+      expect(_isCaretVisible(tester), true);
+
+      await tester.pump(flashPeriod);
```
Updated.

rutvik110 commented:
@matthew-carroll Updated things based on prev review. Major updates are related to

* how we handle the blink stop call from interaction handles as I mentioned [here](https://github.com/superlistapp/super_editor/pull/1778#discussion_r1456402530) and other updates
* Tests are updated to check caret visibility after half caret flash period based on [this](https://github.com/superlistapp/super_editor/pull/1778#discussion_r1452829730) comment.

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_caret_test.dart:
```diff
@@ -166,6 +167,96 @@ void main() {
+
+    testWidgetsOnMobile("does not blink while dragging the caret", (tester) async {
+      // Disable indeterminate animations.
+      BlinkController.indeterminateAnimationsEnabled = false;
```
This line should be unnecessary. Animations are already false at the beginning of every test.

However, the tear down is necessary, so please add a comment explaining why we're adding a teardown.

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_caret_test.dart:
```diff
@@ -166,6 +167,96 @@ void main() {
+      // Drag caret by an arbitrary distance.
+      await tester.dragCaretByDistanceInSuperTextField(const Offset(100, 100));
+
+      // Ensure caret is visible.
```
Add a comment here that explains why the following series of pumps and checks are important. Namely that we don't know exactly how much time has passed, and we might get unlucky and check the visibility when the caret is momentarily visible. Therefore, we test in half-period increments to ensure that we verify the caret isn't blinking at all.

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_caret_test.dart:
```diff
@@ -166,6 +167,96 @@ void main() {
+    });
+
+    testWidgetsOnAndroid("does not blink while dragging collapsed handle", (tester) async {
+      // Disable indeterminate animations.
```
Please apply any relevant comments from the previous test in this one.

matthew-carroll commented on super_text_layout/lib/src/caret_layer.dart:
```diff
@@ -77,7 +77,7 @@ class TextLayoutCaretState extends State<TextLayoutCaret> with TickerProviderSta
 
   BlinkController _createBlinkController() {
```
Let's rename this method from `_createBlinkController()` to `_obtainBlinkController()`

rutvik110 commented:
@matthew-carroll Latest updates,

* Added tests to ensure caret doesn't blink when dragging expanded selection handles in text field.
* Added `dragHandleByDistanceInSuperTextFieldOnMobile` - A top level helper to drag a handle of specific type within text field on mobile.
* Added `_dragAndroidHandleByDistanceInSuperTextField` - Handles android text field handle drag.
* Added `_dragIOSHandleByDistanceInSuperTextField` - Handles iOS text field handle drag.
* Moved `caretBlinkController` dispose call out of schedule frame callback within `SuperAndroidTextField` and `SuperIOSTextField`.
* Other documentation and refactoring updates from last review.

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_caret_test.dart:
```diff
@@ -166,6 +167,182 @@ void main() {
     });
+
+    testWidgetsOnMobile("does not blink while dragging the caret", (tester) async {
+      // Disable indeterminate animations so that pumpAndSettle() doesn't time out.
```
This comment no longer applies here. Please verify that you don't have any other stale comments in this PR.

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_caret_test.dart:
```diff
@@ -166,6 +167,182 @@ void main() {
+      // Configure BlinkController to animate, otherwise it won't blink.
+      BlinkController.indeterminateAnimationsEnabled = true;
+
+      // Drag caret by an arbitrary distance.
```
We should clarify why we're dragging a distance. Readers might assume that this test is focused on what's happening during this drag, but that isn't really the case.

```
// Drag caret by a small distance so that we trigger a user drag event.
// This drag event is continued down below so that we can check for blinking
// during a user drag.
```

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_caret_test.dart:
```diff
@@ -166,6 +167,182 @@ void main() {
+      BlinkController.indeterminateAnimationsEnabled = true;
+
+      // Drag caret by an arbitrary distance.
+      await tester.dragCaretByDistanceInSuperTextField(const Offset(100, 100));
```
We mentioned on the call that you need to get the gesture from this method and then end that gesture at the end of the test.

rutvik110 replied:
Updated.

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_robot.dart:
```diff
@@ -122,6 +123,97 @@ extension SuperTextFieldRobot on WidgetTester {
+  /// Drags the [SuperTextField] handle by the given delta.
+  ///
+  /// {@macro supertextfield_finder}
+  Future<TestGesture> dragHandleByDistanceInSuperTextFieldOnMobile(
```
Can you remind me why you needed to create this? Weren't you already using a method for this when we had our call? Do we not already have dragging behavior in this robot?

rutvik110 replied:
The existing `dragAndroidCollapsedHandleByDistanceInSuperTextField` could only be used for dragging the collapsed handle on `android`. And the earlier tests were testing only for collapsed handle but didn't consider the extended selection handle drag scenario.

After  I added tests for testing the handle drag during the extended text selection, I had to add these methods to support dragging the handles for both android and iOS in that scenario.

matthew-carroll replied:
If we already have a `dragAndroidCollapsedHandleByDistanceInSuperTextField` then rather than add a generic method for any handle type, which creates an ambiguous crossover with the existing method, let's offer explicit handle methods:

`dragUpstreamMobileHandleByDistanceInSuperTextField` and `dragDownstreamMobileHandleByDistanceInSuperTextField`

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_caret_test.dart:
```diff
@@ -166,6 +167,202 @@ void main() {
+
+      final controller = AttributedTextEditingController(
+        text: AttributedText(
+          '''SuperTextField with a content that spans multiple lines of text to test scrolling with  a scrollbar.''',
```
This string doesn't need to use triple quotes. This is a regular single-line string. Please update everywhere in this PR, accordingly.

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_caret_test.dart:
```diff
@@ -166,6 +167,202 @@ void main() {
+
+      // Ensure caret is visible.
+      expect(_isCaretVisible(tester), true);
+    });
```
You still aren't releasing the gesture.

rutvik110 replied:
I added it as a `addTearDown` step after we create the gesture above. 

```
final TestGesture gesture = await tester.dragCaretByDistanceInSuperTextField(const Offset(100, 100));
addTearDown(() => gesture.removePointer());
```

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_caret_test.dart:
```diff
@@ -166,6 +167,202 @@ void main() {
+      // user drag event.
+      // This drag event is continued down below so that we can check for caret blinking
+      // during a user drag.
+      final TestGesture gesture2 =
```
Calling something `2` doesn't communicate anything. Please name the gestures in a way that describes what differentiates them, e.g., `upstreamHandleGesture` and `downstreamHandleGesture`

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_caret_test.dart:
```diff
@@ -166,6 +167,202 @@ void main() {
+
+      // Drag the downstream selection handle by a small distance so that we trigger a
+      // user drag event.
+      // This drag event is continued down below so that we can check for caret blinking
```
Please be mindful of your comment formatting.

If you want multiple sentences together, then the second sentence should immediately follow the first.

If you want two independent sentences, then please add a blank line between them.

rutvik110 replied:
Updated

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_robot.dart:
```diff
@@ -122,6 +123,97 @@ extension SuperTextFieldRobot on WidgetTester {
+    final gesture = await startGesture(handleCenter);
+    await pump(kTapMinTime);
+
+    for (int i = 0; i < 50; i += 1) {
```
Does all of this really need to be repeated? Is there any meaningful difference between teh Android and iOS methods other than the finder for the handle widget?

rutvik110 replied:
No, there isn't. I think a separate `dragMobileHandle` could be created that'll take in the desired finder and use it, while keeping the other logic intact. That way we'll be able to de-dupe this drag steps across similar methods here.

rutvik110 commented:
@matthew-carroll Latest updates, 

* Added explicit methods `dragUpstreamMobileHandleByDistanceInSuperTextField` and `dragDownstreamMobileHandleByDistanceInSuperTextField` to drag upstream or downstream handle on text field.
* Added explicit drag methods for android and ios, 
   * `_dragAndroidUpstreamHandleByDistanceInSuperTextField` 
   * `_dragAndroidDownstreamHandleByDistanceInSuperTextField`
   * `_dragIOSUpstreamHandleByDistanceInSuperTextField`
   * `_dragIOSDownstreamHandleByDistanceInSuperTextField`
* `_dragHandleByDistanceInSuperTextField` added to abstract away the drag logic on the handle.
* Docs updates based on last review.

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_robot.dart:
```diff
@@ -122,6 +123,168 @@ extension SuperTextFieldRobot on WidgetTester {
+    return gesture;
+  }
+
+  /// Drags the [SuperTextField] handle by the given delta.
```
The purpose of Dart Docs is to help people know when/how to use the method. This Dart Doc should include a reference to, and explanation of `superTextFieldHandleFinder`.

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_robot.dart:
```diff
@@ -122,6 +123,168 @@ extension SuperTextFieldRobot on WidgetTester {
+    return gesture;
+  }
+
+  /// Drags the [SuperIOSTextField] downstream handle by the given delta.
```
You should also mention that the method returns the live drag gesture. Please update all such occurrences of docs.

rutvik110 commented:
@matthew-carroll Updated the docs on newly added methods based on last review.

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_robot.dart:
```diff
@@ -122,6 +123,185 @@ extension SuperTextFieldRobot on WidgetTester {
+  ///
+  /// Can be used to drag [SuperTextField] collapsed or expanded selection handles.
+  ///
+  /// Returns the [TestGesture] used to perform the drag.
```
The first sentence should mention what's returned. The first sentence should operate as a summary of the whole method.

It looks like we've had quite a few comments about Dart Docs in this PR. I would encourage you to read through a number of our existing Dart Docs to get a feel for how they should be written so that we can spend much fewer cycles going back and forth about how to write them effectively.

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_robot.dart:
```diff
@@ -122,6 +123,185 @@ extension SuperTextFieldRobot on WidgetTester {
+
+    expect(handleFinder, findsOne);
+
+    final gesture = await _dragHandleByDistanceInSuperTextField(handleFinder, delta);
```
You can return this directly. There's no need to assign it to a variable.

matthew-carroll commented on super_editor/test/super_textfield/super_textfield_robot.dart:
```diff
@@ -122,6 +123,185 @@ extension SuperTextFieldRobot on WidgetTester {
+
+  /// Drags the [SuperIOSTextField] downstream handle by the given delta.
+  ///
+  /// Returns the [TestGesture] used to perform the drag.
```
Same issue here about where the "return" description belongs.

matthew-carroll reviewed (approved):
LGTM
